### PR TITLE
Switch to `nimble_ownership`

### PR DESCRIPTION
**Pull request by whatyouhide**
WIP, we need to publish a new version of nimble_ownership 🙃

### CODE REVIEW

File: lib/mox/server.ex [hunk 1]
```diff
@@ -44,14 +44,16 @@ defmodule Mox.Server do
   # Callbacks
 
   def init(:ok) do
+    {:ok, os} = NimbleOwnership.start_link([])
```
Comment (josevalim) [inline lib/mox/server.ex:47]: Shouldn't we start it on mox supervision tree with a given name? 🤔

File: lib/mox/server.ex [hunk 2]
```diff
@@ -120,18 +124,25 @@ defmodule Mox.Server do
 
   def handle_call(
         {:fetch_fun_to_dispatch, caller_pids, {mock, _, _} = key, source},
-        %{mode: :private, lazy_calls: lazy_calls} = state
+        %{mode: :private} = state
       ) do
-    state = maybe_revalidate_lazy_calls(lazy_calls, state)
+    owner_pid =
+      case NimbleOwnership.get_owner(state.os, caller_pids, mock) do
+        %{owner_pid: owner_pid} -> owner_pid
```
Comment (josevalim) [inline lib/mox/server.ex:131]: Can the owner_pid be nil? If not, shouldn't we merge the conditional below into this function?

File: lib/mox/server.ex [hunk 3]
```diff
@@ -88,17 +90,19 @@ defmodule Mox.Server do
         {:add_expectation, owner_pid, {mock, _, _} = key, expectation},
         %{mode: :private} = state
       ) do
-    if allowance = state.allowances[owner_pid][mock] do
-      {:reply, {:error, {:currently_allowed, allowance}}, state}
-    else
-      state = maybe_add_and_monitor_pid(state, owner_pid)
+    case NimbleOwnership.get_owner(state.os, [owner_pid], mock) do
```
[the remark this answers is not in the document]
Comment (josevalim) [reply lib/mox/server.ex:93]: Or perhaps, we don't even need to support global mode in there. But we would store the expectations in the ownership?

**Review by josevalim — APPROVED** [review]

**Comment by whatyouhide** [comment]
@josevalim done with global mode and no more process in `Mox.Server`. I left the `Mox.Server` module because it's a nice boundary IMO still.

**Comment by josevalim** [comment]
I think we should totally remove the server, if there is no need. :)

**Comment by whatyouhide** [comment]
@josevalim ok, 200 lines less and removed `Mox.Server` completely... worth it 🙃

File: lib/mox.ex [hunk 4]
```diff
@@ -891,4 +910,83 @@ defmodule Mox do
...
+    end
+  end
+
+  defp add_expectation(owner_pid, {mock, _, _} = key, expectation) do
+    # First, make sure that the owner_pid is either the owner or that the mock
+    # isn't owned yet. Otherwise, return an error.
+    case N.fetch_owner(@ownership_server, [owner_pid], mock, @timeout) do
+      {tag, ^owner_pid} when tag in [:ok, :shared_owner] -> :ok
+      {:shared_owner, other_owner} -> throw({:error, {:not_shared_owner, other_owner}})
+      {:ok, other_owner} -> throw({:error, {:currently_allowed, other_owner}})
```
Comment (josevalim) [inline lib/mox.ex:930]: Let's not use throw? If we encapsulate this into a private function, using `with` should be straight-forward.

File: lib/mox.ex [hunk 5]
```diff
@@ -891,4 +910,83 @@ defmodule Mox do
...
+    :ok
+  end
+
+  defp fetch_fun_to_dispatch(caller_pids, {mock, _, _} = key) do
+    # If the mock doesn't have an owner, it can't have expectations so we return :no_expectation.
+    owner_pid =
+      case N.fetch_owner(@ownership_server, caller_pids, mock, @timeout) do
+        {tag, owner_pid} when tag in [:shared_owner, :ok] -> owner_pid
+        :error -> throw(:no_expectation)
+      end
```
Comment (josevalim) [inline lib/mox.ex:959]: Same here.

**Review by josevalim — COMMENTED** [review]
Absolutely beautiful, just two tiny comments. :)

**Pull request merged**